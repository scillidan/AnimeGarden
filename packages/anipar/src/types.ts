export interface ParseOptions {}

export interface ParseResult {
  title: string;

  // Other titles like translations
  titles?: string[];

  // fansub info
  fansub?: {
    // fansub name
    name: string;

    // collab fansubs name
    collab?: string[];
  };

  // season info
  season?: {
    // season number
    number: number;

    // season title
    title?: string;
  };

  // Type, e.g. OVA
  type?: string;

  // episode info
  episode?: EpisodeInfo;

  // Multiple episodes
  episodes?: EpisodeInfo[];

  // episode range
  episodeRange?: {
    from: number;

    to: number;
  };

  // Release version
  version?: string;

  // Video language
  language?: string;

  // Subtitles including language, e.g. 简日内嵌
  subtitles?: string;

  // Source, e.g. WEB-DL
  source?: string;

  // Source platform, e.g. Baha
  platform?: string;

  // Onair year
  year?: number;

  // Onair month
  month?: number;

  // Video file info
  file?: FileInfo;

  // Other tags
  tags: string[];
}

export interface EpisodeInfo {
  number: number;

  // 总集篇 xx.5
  numberSub?: number;

  // Episode title
  title?: string;
}

export interface FileInfo {
  extension?: string;

  audio?: {
    term?: string;
  };

  video?: {
    term?: string;

    resolution?: string;
  };
}
