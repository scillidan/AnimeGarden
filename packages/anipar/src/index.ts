import type { ParseOptions, ParseResult } from './types.ts';

export type { ParseResult };

export function parse(title: string, options: ParseOptions = {}): ParseResult | undefined {
  return undefined;
}
