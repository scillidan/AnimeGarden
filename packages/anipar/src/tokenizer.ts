export class Token {
  public readonly text: string;

  public readonly left: string | undefined;

  public readonly right: string | undefined;

  public constructor(text: string, left?: string, right?: string) {
    this.text = text;
    this.left = left;
    this.right = right;
  }

  public get isWrapped() {
    return this.left && this.right;
  }

  public slice(start: number, end?: number) {
    const text = this.text.slice(start, end);
    return new Token(text, this.left, this.right);
  }

  public trim() {
    const text = this.text.trim();
    return new Token(text, this.left, this.right);
  }

  public toString() {
    return `${this.left ?? ''}${this.text}${this.right ?? ''}`;
  }
}

export const Wrappers = new Map([
  ['[', ']'],
  ['【', '】'],
  ['(', ')'],
  ['（', '）'],
  ['{', '}']
]);

export const RevWrappers = new Map([...Wrappers.entries()].map(([k, v]) => [v, k]));

export function tokenize(text: string) {
  const tokens: Token[] = [];

  let cursor = 0;
  let cur = '';
  let left: string | undefined = undefined;
  let right: string | undefined = undefined;
  while (cursor < text.length) {
    const char = text[cursor];
    if (Wrappers.has(char)) {
      if (cur) {
        tokens.push(new Token(cur.trim()));
        cur = '';
      }
      left = char;
      right = Wrappers.get(char)!;
    } else if (left && right && char === right) {
      tokens.push(new Token(cur.trim(), left, right));
      cur = '';
      left = undefined;
      right = undefined;
    } else {
      cur += char;
    }
    cursor += 1;
  }
  if (cur) {
    tokens.push(new Token(cur.trim()));
  }

  return tokens.filter((t) => t.text);
}
