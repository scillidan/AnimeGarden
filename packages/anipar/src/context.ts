import type { ParseOptions, ParseResult } from './types';

import { Token } from './tokenizer';

export class Context {
  // Left consumed tags
  public left = 0;

  // Right consumed tags
  public right = 0;

  public tokens: Token[];

  public options: ParseOptions;

  public result: Partial<ParseResult> = {};

  public tags: string[] = [];

  public constructor(tokens: Token[], options: ParseOptions) {
    this.tokens = tokens;
    this.options = options;
  }

  public update<K1 extends keyof ParseResult>(key: K1, value: ParseResult[K1]) {
    this.result[key] = value;
  }

  public update2<
    K1 extends keyof Required<ParseResult>,
    K2 extends keyof Required<ParseResult>[K1]
  >(key1: K1, key2: K2, value: Required<ParseResult>[K1][K2]) {
    if (!this.result[key1]) {
      // @ts-expect-error
      this.result[key1] = {};
    }
    // @ts-expect-error
    this.result[key1][key2] = value;
  }

  public validate(): ParseResult | undefined {
    return this.result as ParseResult;
  }
}
