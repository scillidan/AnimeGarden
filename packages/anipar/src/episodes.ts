import { Context } from './context';

const WrappedEpisodeRE = /^(?<ep1>\d+)(?:\.(\d))?(?:[vV](\d+))?$|^第(?<ep2>\d+)[集话話]$/;
const WrappedMovieRE = /^Movie [vV](\d+)$/;

const EpisodesRange1 = /^(\d+)-(\d+)(?:\s*.*)$/;
const EpisodesRange2 = /^全(\d+)集$/;

export function parseWrappedEpisodes(ctx: Context) {
  const token = ctx.tokens[ctx.right];
  const text = token.text;
  if (token.isWrapped && matchEpiodes(ctx, text)) {
    ctx.right -= 1;
    return true;
  }
  return false;
}

export function matchEpiodes(ctx: Context, text: string) {
  // 1. Single episode
  {
    const res = WrappedEpisodeRE.exec(text);
    if (res) {
      const epText = res.groups?.ep1! || res.groups?.ep2!;
      const ep = +epText;
      if (!Number.isNaN(ep)) {
        // Handle year: [2024]
        if (1949 <= ep && ep <= 2099 && text === epText && ctx.hasEpisode) {
          ctx.update('year', ep);
          return true;
        }

        ctx.update2('episode', 'number', ep);
        // 1.5
        if (res[2] && !Number.isNaN(res[2])) {
          ctx.update2('episode', 'numberSub', +res[2]);
        }
        // 3v2
        if (res[3] && !Number.isNaN(res[3])) {
          ctx.update('version', +res[3]);
        }
        return true;
      }
    }
  }
  {
    // [Movie] / [Movie v2]
    const res = WrappedMovieRE.exec(text);
    if (res) {
      if (res[1] && !Number.isNaN(res[1])) {
        ctx.update('version', +res[1]);
      }
      ctx.update('type', 'Movie');
      return true;
    }
  }

  // 2. Episode range
  {
    // 01-26
    const res = EpisodesRange1.exec(text);
    if (res) {
      const from = +res[1];
      const to = +res[2];
      if (!Number.isNaN(from) && !Number.isNaN(to)) {
        ctx.update2('episodeRange', 'from', from);
        ctx.update2('episodeRange', 'to', to);
        return true;
      }
    }
  }
  {
    // 全26集
    const res = EpisodesRange2.exec(text);
    if (res) {
      const to = +res[1];
      if (!Number.isNaN(to)) {
        ctx.update2('episodeRange', 'from', 1);
        ctx.update2('episodeRange', 'to', to);
        return true;
      }
    }
  }

  return false;
}

const SuffixEpisodeRE = [/- (\d+)(?:\.(\d))?$/, /第(\d+)(?:\.(\d))?[集话話]$/];

export function parseSuffixEpisodes(ctx: Context) {
  if (ctx.result.episode || ctx.result.episodes || ctx.result.episodeRange) {
    return true;
  }

  const token = ctx.tokens[ctx.right];
  const text = token.text;
  for (const RE of SuffixEpisodeRE) {
    const res = RE.exec(text);
    if (res) {
      const ep = +res[1];
      if (!Number.isNaN(ep)) {
        ctx.update2('episode', 'number', ep);

        // 1.5
        if (res[2] && !Number.isNaN(res[2])) {
          ctx.update2('episode', 'numberSub', +res[2]);
        }

        // Remove token suffix
        ctx.tokens[ctx.right] = token.slice(0, text.length - res[0].length).trim();

        return true;
      }
    }
  }
}

// Season regexp
export const SuffixSeasonOrEpisodesRes: Array<
  [RegExp, (res: RegExpExecArray, ctx: Context) => boolean]
> = [
  [
    /(?:S|Season\s?)(\d+)$/,
    (res, ctx) => {
      const season = +res[1];
      if (!Number.isNaN(season)) {
        if (!ctx.result.season?.number || ctx.result.season.number === season) {
          ctx.update2('season', 'number', season);
          return true;
        }
      }
      return false;
    }
  ],
  [
    /(1st|2nd|3rd|[456789]th) Season$/,
    (res, ctx) => {
      const season = Number.parseInt(res[1]);
      if (!Number.isNaN(season)) {
        if (!ctx.result.season?.number || ctx.result.season.number === season) {
          ctx.update2('season', 'number', season);
          return true;
        }
      }
      return false;
    }
  ],
  [
    /第?(\d+)[季期]$/,
    (res, ctx) => {
      const season = +res[1];
      if (!Number.isNaN(season)) {
        if (!ctx.result.season?.number || ctx.result.season.number === season) {
          ctx.update2('season', 'number', season);
          return true;
        }
      }
      return false;
    }
  ],
  [
    /第?((?:[零一二三四五六七八九]十)?[零一二三四五六七八九])[季期]$/,
    (res, ctx) => {
      const text = res[1];
      const map = {
        零: 0,
        一: 1,
        二: 2,
        三: 3,
        四: 4,
        五: 5,
        六: 6,
        七: 7,
        八: 8,
        九: 9
      };
      const base =
        text.length === 2 && text[0] === '十'
          ? 1
          : text.length === 3
            ? map[text[0] as keyof typeof map]!
            : 0;
      const offset = map[text[text.length - 1] as keyof typeof map];
      const season = base * 10 + offset;
      if (!ctx.result.season?.number || ctx.result.season.number === season) {
        ctx.update2('season', 'number', season);
        return true;
      }
      return false;
    }
  ],
  [
    / (\d+)$/,
    (res, ctx) => {
      const season = +res[1];
      if (
        !Number.isNaN(season) &&
        (ctx.result.episode || ctx.result.episodes || ctx.result.episodeRange)
      ) {
        // Skip: Part 2
        if (/Parts? \d+$/.test(res.input)) {
          return false;
        }
        if (!ctx.result.season?.number || ctx.result.season.number === season) {
          ctx.update2('season', 'number', season);
          return true;
        }
      }
      return false;
    }
  ],
  [
    /(?<ep1>\d+)(?:\.(\d))?(?:[vV](\d+))?$|^第(?<ep2>\d+)[集话話]$/,
    (res, ctx) => {
      const ep = +(res.groups?.ep1! || res.groups?.ep2!);
      if (!Number.isNaN(ep)) {
        // Skip: Part 2
        if (/Parts? \d+$/.test(res.input)) {
          return false;
        }

        ctx.update2('episode', 'number', ep);
        // 1.5
        if (res[2] && !Number.isNaN(res[2])) {
          ctx.update2('episode', 'numberSub', +res[2]);
        }
        // 3v2
        if (res[3] && !Number.isNaN(res[3])) {
          ctx.update('version', +res[3]);
        }
        return true;
      }
      return false;
    }
  ],
  [
    /（日配版）$/,
    (res, ctx) => {
      ctx.update('language', res[0]);
      return true;
    }
  ],
  [
    /（重播）$/,
    (res, ctx) => {
      ctx.tags.push(res[0]);
      return true;
    }
  ]
];
