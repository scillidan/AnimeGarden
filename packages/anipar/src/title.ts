import { Context } from './context';
import { parseSuffixEpisodes, parseSuffixSeasonOrEpisodes } from './episodes';

export function parseFansub(ctx: Context) {
  // [fansub] title
  if (ctx.left + 1 > ctx.right) {
    return false;
  }

  const token = ctx.tokens[ctx.left];
  if (token.isWrapped) {
    const text = token.text;

    let found = false;
    const seps = ['&', '·'];
    for (const sep of seps) {
      const [name, ...collab] = text.split(sep);

      if (collab.length > 0) {
        // Hack: [jibaketa合成&壓制]
        if (collab.length === 1 && collab[0] === '壓制') {
          break;
        }

        found = true;
        ctx.update2('fansub', 'name', name);
        ctx.update2('fansub', 'collab', collab);
        break;
      }
    }

    if (!found) {
      ctx.update2('fansub', 'name', text);
    }

    ctx.left += 1;

    return true;
  }

  return false;
}

export function parseTitle(ctx: Context) {
  // 1. Parse suffix episodes
  parseSuffixEpisodes(ctx);

  // 2. Parse titles
  const rest = ctx.tokens.slice(ctx.left, ctx.right + 1);
  if (rest.length === 0) return false;

  const text = rest.length === 1 ? rest[0].text : rest.map((t) => t.toString()).join('');

  // 2.1. Try split multiple titles
  let found = false;
  const separators = ['/', '\\'];
  for (const sep of separators) {
    const parts = text
      .split(sep)
      .map((t) => t.trim())
      .filter((t) => !!t);
    if (parts.length > 1) {
      const [title, ...other] = parts;

      const trimmedTitle = parseSuffixSeasonOrEpisodes(ctx, title);
      const trimmedOther = other
        .map((t) => parseSuffixSeasonOrEpisodes(ctx, t))
        .map((t) => t.trim())
        .filter((t) => !!t);

      ctx.update('title', trimmedTitle);
      if (trimmedOther.length > 0) {
        ctx.update('titles', trimmedOther);
      }
      found = true;
    }
  }
  if (!found) {
    const trimmed = parseSuffixSeasonOrEpisodes(ctx, text);
    ctx.update('title', trimmed);
  }

  return true;
}
