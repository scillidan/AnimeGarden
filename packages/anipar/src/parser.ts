import type { ParseOptions, ParseResult } from './types';

import { Context } from './context';
import { tokenize } from './tokenizer';
import { parseLeftTags, parseRightTags } from './keyword';
import { parseFansub, parseTitle } from './title';

export function parse(title: string, options: ParseOptions = {}): ParseResult | undefined {
  const tokens = tokenize(title);
  if (tokens.length === 0) return undefined;

  const context = new Context(tokens, options);

  // 1. Parse right tags
  parseRightTags(context);
  // 2. Parse left tags
  parseLeftTags(context);
  // 3. Parse fansub
  parseFansub(context);
  parseLeftTags(context);
  // 4. Parse title
  if (!parseTitle(context)) {
    return undefined;
  }

  // 6. Return result
  return context.validate();
}
