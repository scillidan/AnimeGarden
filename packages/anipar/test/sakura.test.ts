import { describe, it, expect } from 'vitest';

import { parse } from '../src';

const titles = [
  '[桜都字幕组] 超自然武装当哒当 / Dan Da Dan [11][1080p][简繁内封]',
  '[桜都字幕組] 膽大黨 / Dan Da Dan [11][1080p][繁體內嵌]',
  '[桜都字幕组] 超自然武装当哒当 / Dan Da Dan [11][1080p][简体内嵌]',
  '[桜都字幕組] 常軌脫離Creative / Hamidashi Creative [11][1080p][繁體內嵌]',
  '[桜都字幕组] 常轨脱离Creative / Hamidashi Creative [11][1080p][简繁内封]',
  '[桜都字幕组] 常轨脱离Creative / Hamidashi Creative [11][1080p][简体内嵌]',
  '[桜都字幕组] 在地下城寻求邂逅是否搞错了什么 第五季 / Dungeon ni Deai o Motomeru no wa Machigatte Iru Darouka： Familia Myth V [10][1080p][简繁内封][v2]',
  '[桜都字幕組] 在地下城尋求邂逅是否搞錯了什麼 第五季 / Dungeon ni Deai o Motomeru no wa Machigatte Iru Darouka： Familia Myth V  [10][1080P][繁體內嵌][v2]',
  '[桜都字幕组] 在地下城寻求邂逅是否搞错了什么 第五季 / Dungeon ni Deai o Motomeru no wa Machigatte Iru Darouka： Familia Myth V [10][1080p][简体内嵌][v2]',
  '[桜都字幕组] 在地下城寻求邂逅是否搞错了什么 第五季 / Dungeon ni Deai o Motomeru no wa Machigatte Iru Darouka： Familia Myth V [10][1080p][简体内嵌]',
  '[桜都字幕組] 在地下城尋求邂逅是否搞錯了什麼 第五季 / Dungeon ni Deai o Motomeru no wa Machigatte Iru Darouka： Familia Myth V [10][1080P][繁體內嵌]',
  '[桜都字幕组] 在地下城寻求邂逅是否搞错了什么 第五季 / Dungeon ni Deai o Motomeru no wa Machigatte Iru Darouka： Familia Myth V [10][1080p][简繁内封]',
  '[桜都字幕組] BLEACH 死神 千年血戰篇 -相剋譚- / Bleach：Sennen Kessen Hen - Soukoku Tan [10][1080p][繁體內嵌]',
  '[桜都字幕组] BLEACH 死神 千年血战篇 -相克谭- / Bleach：Sennen Kessen Hen - Soukoku Tan [10][1080p][简体内嵌]',
  '[桜都字幕组] BLEACH 死神 千年血战篇 -相克谭- / Bleach：Sennen Kessen Hen - Soukoku Tan [10][1080p][简繁内封]',
  '[桜都字幕組] 大正偽婚～替身新娘與軍服的猛愛 / Taishou Itsuwari Bridal： Migawari Hanayome to Gunpuku no Mou Ai [08][1080p][繁體內嵌]',
  '[桜都字幕组] 大正伪婚～替身新娘与军服的猛爱 / Taishou Itsuwari Bridal： Migawari Hanayome to Gunpuku no Mou Ai [08][1080p][简体内嵌]',
  '[桜都字幕组] 大正伪婚～替身新娘与军服的猛爱 / Taishou Itsuwari Bridal： Migawari Hanayome to Gunpuku no Mou Ai [08][1080p][简繁内封]',
  '[桜都字幕组] 亦叶亦花 / Nanare Hananare [01-12END][1080p][简体内嵌]',
  '[桜都字幕组] 亦叶亦花 / Nanare Hananare [01-12][1080p][简繁内封]',
  '[桜都字幕組] 少女如草花綻放 / Nanare Hananare [01-12][1080p][繁體內嵌]',
  '[桜都字幕組] 妻子变成小学生。 / Tsuma, Shougakusei ni Naru  [11][1080p][简体內嵌]',
  '[桜都字幕組] 妻子變成小學生。 / Tsuma, Shougakusei ni Naru  [11][1080p][繁体内嵌]',
  '[桜都字幕組] 妻子变成小学生。 / Tsuma, Shougakusei ni Naru  [11][1080p][简繁内封]',
  '[桜都字幕组] 在地下城寻求邂逅是否搞错了什么 第五季 / Dungeon ni Deai o Motomeru no wa Machigatte Iru Darouka： Familia Myth V [09][1080p][简体内嵌]'
];

describe('sakura', () => {
  for (const title of titles) {
    it(title, () => {
      expect(parse(title)).toMatchSnapshot();
    });
  }
});
