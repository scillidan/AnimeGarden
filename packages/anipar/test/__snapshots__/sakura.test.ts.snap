// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`sakura > [桜都字幕組] BLEACH 死神 千年血戰篇 -相剋譚- / Bleach：Sennen Kessen Hen - Soukoku Tan [10][1080p][繁體內嵌] 1`] = `
{
  "episode": {
    "number": 10,
  },
  "fansub": {
    "name": "桜都字幕組",
  },
  "file": {
    "video": {
      "resolution": "1080p",
    },
  },
  "language": "繁體",
  "subtitles": "內嵌",
  "tags": [],
  "title": "BLEACH 死神 千年血戰篇 -相剋譚-",
  "titles": [
    "Bleach：Sennen Kessen Hen - Soukoku Tan",
  ],
}
`;

exports[`sakura > [桜都字幕組] 在地下城尋求邂逅是否搞錯了什麼 第五季 / Dungeon ni Deai o Motomeru no wa Machigatte Iru Darouka： Familia Myth V  [10][1080P][繁體內嵌][v2] 1`] = `
{
  "episode": {
    "number": 10,
  },
  "fansub": {
    "name": "桜都字幕組",
  },
  "file": {
    "video": {
      "resolution": "1080P",
    },
  },
  "language": "繁體",
  "season": {
    "number": 5,
  },
  "subtitles": "內嵌",
  "tags": [],
  "title": "在地下城尋求邂逅是否搞錯了什麼",
  "titles": [
    "Dungeon ni Deai o Motomeru no wa Machigatte Iru Darouka： Familia Myth V",
  ],
  "version": 2,
}
`;

exports[`sakura > [桜都字幕組] 在地下城尋求邂逅是否搞錯了什麼 第五季 / Dungeon ni Deai o Motomeru no wa Machigatte Iru Darouka： Familia Myth V [10][1080P][繁體內嵌] 1`] = `
{
  "episode": {
    "number": 10,
  },
  "fansub": {
    "name": "桜都字幕組",
  },
  "file": {
    "video": {
      "resolution": "1080P",
    },
  },
  "language": "繁體",
  "season": {
    "number": 5,
  },
  "subtitles": "內嵌",
  "tags": [],
  "title": "在地下城尋求邂逅是否搞錯了什麼",
  "titles": [
    "Dungeon ni Deai o Motomeru no wa Machigatte Iru Darouka： Familia Myth V",
  ],
}
`;

exports[`sakura > [桜都字幕組] 大正偽婚～替身新娘與軍服的猛愛 / Taishou Itsuwari Bridal： Migawari Hanayome to Gunpuku no Mou Ai [08][1080p][繁體內嵌] 1`] = `
{
  "episode": {
    "number": 8,
  },
  "fansub": {
    "name": "桜都字幕組",
  },
  "file": {
    "video": {
      "resolution": "1080p",
    },
  },
  "language": "繁體",
  "subtitles": "內嵌",
  "tags": [],
  "title": "大正偽婚～替身新娘與軍服的猛愛",
  "titles": [
    "Taishou Itsuwari Bridal： Migawari Hanayome to Gunpuku no Mou Ai",
  ],
}
`;

exports[`sakura > [桜都字幕組] 妻子变成小学生。 / Tsuma, Shougakusei ni Naru  [11][1080p][简体內嵌] 1`] = `
{
  "episode": {
    "number": 11,
  },
  "fansub": {
    "name": "桜都字幕組",
  },
  "file": {
    "video": {
      "resolution": "1080p",
    },
  },
  "language": "简体",
  "subtitles": "內嵌",
  "tags": [],
  "title": "妻子变成小学生。",
  "titles": [
    "Tsuma, Shougakusei ni Naru",
  ],
}
`;

exports[`sakura > [桜都字幕組] 妻子变成小学生。 / Tsuma, Shougakusei ni Naru  [11][1080p][简繁内封] 1`] = `
{
  "episode": {
    "number": 11,
  },
  "fansub": {
    "name": "桜都字幕組",
  },
  "file": {
    "video": {
      "resolution": "1080p",
    },
  },
  "language": "简繁",
  "subtitles": "内封",
  "tags": [],
  "title": "妻子变成小学生。",
  "titles": [
    "Tsuma, Shougakusei ni Naru",
  ],
}
`;

exports[`sakura > [桜都字幕組] 妻子變成小學生。 / Tsuma, Shougakusei ni Naru  [11][1080p][繁体内嵌] 1`] = `
{
  "episode": {
    "number": 11,
  },
  "fansub": {
    "name": "桜都字幕組",
  },
  "file": {
    "video": {
      "resolution": "1080p",
    },
  },
  "tags": [
    "繁体内嵌",
  ],
  "title": "妻子變成小學生。",
  "titles": [
    "Tsuma, Shougakusei ni Naru",
  ],
}
`;

exports[`sakura > [桜都字幕組] 少女如草花綻放 / Nanare Hananare [01-12][1080p][繁體內嵌] 1`] = `
{
  "episodeRange": {
    "from": 1,
    "to": 12,
  },
  "fansub": {
    "name": "桜都字幕組",
  },
  "file": {
    "video": {
      "resolution": "1080p",
    },
  },
  "language": "繁體",
  "subtitles": "內嵌",
  "tags": [],
  "title": "少女如草花綻放",
  "titles": [
    "Nanare Hananare",
  ],
}
`;

exports[`sakura > [桜都字幕組] 常軌脫離Creative / Hamidashi Creative [11][1080p][繁體內嵌] 1`] = `
{
  "episode": {
    "number": 11,
  },
  "fansub": {
    "name": "桜都字幕組",
  },
  "file": {
    "video": {
      "resolution": "1080p",
    },
  },
  "language": "繁體",
  "subtitles": "內嵌",
  "tags": [],
  "title": "常軌脫離Creative",
  "titles": [
    "Hamidashi Creative",
  ],
}
`;

exports[`sakura > [桜都字幕組] 膽大黨 / Dan Da Dan [11][1080p][繁體內嵌] 1`] = `
{
  "episode": {
    "number": 11,
  },
  "fansub": {
    "name": "桜都字幕組",
  },
  "file": {
    "video": {
      "resolution": "1080p",
    },
  },
  "language": "繁體",
  "subtitles": "內嵌",
  "tags": [],
  "title": "膽大黨",
  "titles": [
    "Dan Da Dan",
  ],
}
`;

exports[`sakura > [桜都字幕组] BLEACH 死神 千年血战篇 -相克谭- / Bleach：Sennen Kessen Hen - Soukoku Tan [10][1080p][简体内嵌] 1`] = `
{
  "episode": {
    "number": 10,
  },
  "fansub": {
    "name": "桜都字幕组",
  },
  "file": {
    "video": {
      "resolution": "1080p",
    },
  },
  "language": "简体",
  "subtitles": "内嵌",
  "tags": [],
  "title": "BLEACH 死神 千年血战篇 -相克谭-",
  "titles": [
    "Bleach：Sennen Kessen Hen - Soukoku Tan",
  ],
}
`;

exports[`sakura > [桜都字幕组] BLEACH 死神 千年血战篇 -相克谭- / Bleach：Sennen Kessen Hen - Soukoku Tan [10][1080p][简繁内封] 1`] = `
{
  "episode": {
    "number": 10,
  },
  "fansub": {
    "name": "桜都字幕组",
  },
  "file": {
    "video": {
      "resolution": "1080p",
    },
  },
  "language": "简繁",
  "subtitles": "内封",
  "tags": [],
  "title": "BLEACH 死神 千年血战篇 -相克谭-",
  "titles": [
    "Bleach：Sennen Kessen Hen - Soukoku Tan",
  ],
}
`;

exports[`sakura > [桜都字幕组] 亦叶亦花 / Nanare Hananare [01-12][1080p][简繁内封] 1`] = `
{
  "episodeRange": {
    "from": 1,
    "to": 12,
  },
  "fansub": {
    "name": "桜都字幕组",
  },
  "file": {
    "video": {
      "resolution": "1080p",
    },
  },
  "language": "简繁",
  "subtitles": "内封",
  "tags": [],
  "title": "亦叶亦花",
  "titles": [
    "Nanare Hananare",
  ],
}
`;

exports[`sakura > [桜都字幕组] 亦叶亦花 / Nanare Hananare [01-12END][1080p][简体内嵌] 1`] = `
{
  "episodeRange": {
    "from": 1,
    "to": 12,
  },
  "fansub": {
    "name": "桜都字幕组",
  },
  "file": {
    "video": {
      "resolution": "1080p",
    },
  },
  "language": "简体",
  "subtitles": "内嵌",
  "tags": [],
  "title": "亦叶亦花",
  "titles": [
    "Nanare Hananare",
  ],
}
`;

exports[`sakura > [桜都字幕组] 在地下城寻求邂逅是否搞错了什么 第五季 / Dungeon ni Deai o Motomeru no wa Machigatte Iru Darouka： Familia Myth V [09][1080p][简体内嵌] 1`] = `
{
  "episode": {
    "number": 9,
  },
  "fansub": {
    "name": "桜都字幕组",
  },
  "file": {
    "video": {
      "resolution": "1080p",
    },
  },
  "language": "简体",
  "season": {
    "number": 5,
  },
  "subtitles": "内嵌",
  "tags": [],
  "title": "在地下城寻求邂逅是否搞错了什么",
  "titles": [
    "Dungeon ni Deai o Motomeru no wa Machigatte Iru Darouka： Familia Myth V",
  ],
}
`;

exports[`sakura > [桜都字幕组] 在地下城寻求邂逅是否搞错了什么 第五季 / Dungeon ni Deai o Motomeru no wa Machigatte Iru Darouka： Familia Myth V [10][1080p][简体内嵌] 1`] = `
{
  "episode": {
    "number": 10,
  },
  "fansub": {
    "name": "桜都字幕组",
  },
  "file": {
    "video": {
      "resolution": "1080p",
    },
  },
  "language": "简体",
  "season": {
    "number": 5,
  },
  "subtitles": "内嵌",
  "tags": [],
  "title": "在地下城寻求邂逅是否搞错了什么",
  "titles": [
    "Dungeon ni Deai o Motomeru no wa Machigatte Iru Darouka： Familia Myth V",
  ],
}
`;

exports[`sakura > [桜都字幕组] 在地下城寻求邂逅是否搞错了什么 第五季 / Dungeon ni Deai o Motomeru no wa Machigatte Iru Darouka： Familia Myth V [10][1080p][简体内嵌][v2] 1`] = `
{
  "episode": {
    "number": 10,
  },
  "fansub": {
    "name": "桜都字幕组",
  },
  "file": {
    "video": {
      "resolution": "1080p",
    },
  },
  "language": "简体",
  "season": {
    "number": 5,
  },
  "subtitles": "内嵌",
  "tags": [],
  "title": "在地下城寻求邂逅是否搞错了什么",
  "titles": [
    "Dungeon ni Deai o Motomeru no wa Machigatte Iru Darouka： Familia Myth V",
  ],
  "version": 2,
}
`;

exports[`sakura > [桜都字幕组] 在地下城寻求邂逅是否搞错了什么 第五季 / Dungeon ni Deai o Motomeru no wa Machigatte Iru Darouka： Familia Myth V [10][1080p][简繁内封] 1`] = `
{
  "episode": {
    "number": 10,
  },
  "fansub": {
    "name": "桜都字幕组",
  },
  "file": {
    "video": {
      "resolution": "1080p",
    },
  },
  "language": "简繁",
  "season": {
    "number": 5,
  },
  "subtitles": "内封",
  "tags": [],
  "title": "在地下城寻求邂逅是否搞错了什么",
  "titles": [
    "Dungeon ni Deai o Motomeru no wa Machigatte Iru Darouka： Familia Myth V",
  ],
}
`;

exports[`sakura > [桜都字幕组] 在地下城寻求邂逅是否搞错了什么 第五季 / Dungeon ni Deai o Motomeru no wa Machigatte Iru Darouka： Familia Myth V [10][1080p][简繁内封][v2] 1`] = `
{
  "episode": {
    "number": 10,
  },
  "fansub": {
    "name": "桜都字幕组",
  },
  "file": {
    "video": {
      "resolution": "1080p",
    },
  },
  "language": "简繁",
  "season": {
    "number": 5,
  },
  "subtitles": "内封",
  "tags": [],
  "title": "在地下城寻求邂逅是否搞错了什么",
  "titles": [
    "Dungeon ni Deai o Motomeru no wa Machigatte Iru Darouka： Familia Myth V",
  ],
  "version": 2,
}
`;

exports[`sakura > [桜都字幕组] 大正伪婚～替身新娘与军服的猛爱 / Taishou Itsuwari Bridal： Migawari Hanayome to Gunpuku no Mou Ai [08][1080p][简体内嵌] 1`] = `
{
  "episode": {
    "number": 8,
  },
  "fansub": {
    "name": "桜都字幕组",
  },
  "file": {
    "video": {
      "resolution": "1080p",
    },
  },
  "language": "简体",
  "subtitles": "内嵌",
  "tags": [],
  "title": "大正伪婚～替身新娘与军服的猛爱",
  "titles": [
    "Taishou Itsuwari Bridal： Migawari Hanayome to Gunpuku no Mou Ai",
  ],
}
`;

exports[`sakura > [桜都字幕组] 大正伪婚～替身新娘与军服的猛爱 / Taishou Itsuwari Bridal： Migawari Hanayome to Gunpuku no Mou Ai [08][1080p][简繁内封] 1`] = `
{
  "episode": {
    "number": 8,
  },
  "fansub": {
    "name": "桜都字幕组",
  },
  "file": {
    "video": {
      "resolution": "1080p",
    },
  },
  "language": "简繁",
  "subtitles": "内封",
  "tags": [],
  "title": "大正伪婚～替身新娘与军服的猛爱",
  "titles": [
    "Taishou Itsuwari Bridal： Migawari Hanayome to Gunpuku no Mou Ai",
  ],
}
`;

exports[`sakura > [桜都字幕组] 常轨脱离Creative / Hamidashi Creative [11][1080p][简体内嵌] 1`] = `
{
  "episode": {
    "number": 11,
  },
  "fansub": {
    "name": "桜都字幕组",
  },
  "file": {
    "video": {
      "resolution": "1080p",
    },
  },
  "language": "简体",
  "subtitles": "内嵌",
  "tags": [],
  "title": "常轨脱离Creative",
  "titles": [
    "Hamidashi Creative",
  ],
}
`;

exports[`sakura > [桜都字幕组] 常轨脱离Creative / Hamidashi Creative [11][1080p][简繁内封] 1`] = `
{
  "episode": {
    "number": 11,
  },
  "fansub": {
    "name": "桜都字幕组",
  },
  "file": {
    "video": {
      "resolution": "1080p",
    },
  },
  "language": "简繁",
  "subtitles": "内封",
  "tags": [],
  "title": "常轨脱离Creative",
  "titles": [
    "Hamidashi Creative",
  ],
}
`;

exports[`sakura > [桜都字幕组] 超自然武装当哒当 / Dan Da Dan [11][1080p][简体内嵌] 1`] = `
{
  "episode": {
    "number": 11,
  },
  "fansub": {
    "name": "桜都字幕组",
  },
  "file": {
    "video": {
      "resolution": "1080p",
    },
  },
  "language": "简体",
  "subtitles": "内嵌",
  "tags": [],
  "title": "超自然武装当哒当",
  "titles": [
    "Dan Da Dan",
  ],
}
`;

exports[`sakura > [桜都字幕组] 超自然武装当哒当 / Dan Da Dan [11][1080p][简繁内封] 1`] = `
{
  "episode": {
    "number": 11,
  },
  "fansub": {
    "name": "桜都字幕组",
  },
  "file": {
    "video": {
      "resolution": "1080p",
    },
  },
  "language": "简繁",
  "subtitles": "内封",
  "tags": [],
  "title": "超自然武装当哒当",
  "titles": [
    "Dan Da Dan",
  ],
}
`;
